transaction/builder/src/memo_builder/rth_memo_builder.rs

```diff
@@ -82,17 +82,41 @@ pub enum CustomMemoType {
     FlexibleMemoGenerator(FlexibleMemoGenerator),
 }
 
-/// A function that returns a flexible memopayload for outputs
-pub type FlexibleOutputMemoGenerator = Box<
-    dyn Fn(FlexibleMemoOutputContext) -> Result<FlexibleMemoPayload, NewMemoError> + Sync + Send,
->;
-/// A function that returns a flexible memopayload for change
-pub type FlexibleChangeMemoGenerator = Box<
-    dyn Fn(FlexibleMemoChangeContext) -> Result<FlexibleMemoPayload, NewMemoError> + Sync + Send,
->;
+/// This is a trait that must be implemented by generators used for creating
+/// flexible output memos and flexible change memos
+pub trait FlexibleMemoGeneratorTrait: Sync + Send + Debug {
+    /// This is called when generating output memos.
+    /// It must return a memo payload with the first byte being 0x01
+    fn create_output_memo(
+        &self,
+        context: FlexibleMemoOutputContext,
+    ) -> Result<FlexibleMemoPayload, NewMemoError>;
+    /// This is called when generating change memos.
+    /// It must return a memo payload with the first byte being 0x02
+    fn create_change_memo(
+        &self,
+        context: FlexibleMemoChangeContext,
+    ) -> Result<FlexibleMemoPayload, NewMemoError>;
+}
+
+/// This is a struct that contains a generator used for creating flexible output
+/// memos and flexible change memos
+#[derive(Clone)]
+pub struct FlexibleMemoGenerator {
+    /// This is a reference to the actual generator that will be called.
+    pub generator: Arc<Box<dyn FlexibleMemoGeneratorTrait>>,
+}
+
+impl Debug for FlexibleMemoGenerator {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        f.debug_struct("FlexibleMemoGenerator")
+            .field("generator", &self.generator)
+            .finish()
+    }
+}
```

Comment:
Since we're removing this struct, we'll want to implement `Debug` for the concrete struct (`FlexibleMemoGeneratorPaymentRequest`) that implement  the trait. So I recommend removing this here and adding it below.